src/runtime/graph/graph_runtime.cc

```diff
@@ -96,6 +96,44 @@ void GraphRuntime::SetInput(int index, DLTensor* data_in) {
   uint32_t eid = this->entry_id(input_nodes_[index], 0);
   data_entry_[eid].CopyFrom(data_in);
 }
+/*!
+ * \brief set index-th input to the graph without copying the data.
+ * \param index The input index.
+ * \param data_ref The input data that is referred.
+ */
+void GraphRuntime::SetInputZeroCopy(int index, DLTensor* data_ref) {
+  CHECK_LT(static_cast<size_t>(index), input_nodes_.size());
+  uint32_t eid = this->entry_id(input_nodes_[index], 0);
+  const auto& shape = dltensor_entry_shapes_[eid];
+
+  // check the consistency of input shape
+  CHECK_EQ(data_alignment_[eid], GetDataAlignment(*data_ref));
+  CHECK(reinterpret_cast<size_t>(data_ref->data) % kAllocAlignment == 0);
+  if (shape.size() == static_cast<size_t>(data_ref->ndim)) {
+    for (auto i = 0; i < data_ref->ndim; ++i) {
+      CHECK_EQ(shape[i], data_ref->shape[i]);
+    }
+  } else {
+    int64_t acc_prev =
+        std::accumulate(shape.data(), shape.data() + shape.size(), 1, std::multiplies<int64_t>());
+    int64_t acc = std::accumulate(data_ref->shape, data_ref->shape + data_ref->ndim, 1,
+                                  std::multiplies<int64_t>());
+    CHECK_EQ(acc_prev, acc);
+  }
+
+  // Update the data pointer for each argument of each op
+  for (auto& op_arg : op_args_) {
```

Comment:
I'm thinking, maybe we can change the API to accept an array of DLTensors, update the entries in the data_entries_ for all of them, and then call `SetupOpExecs()`. That way, you don't need to save the `input_entry_ids` in `op_args_` and the code can be much cleaner.

Reply:
I had an implementation like this but it turned out the be more difficult. 
1. To update `data_entries_`, you need to create a DLManagedTensor, which is more small allocation.  
2. Running `SetupOpExecs()` seems a bit heavy.